thingsearch/service/src/main/java/org/eclipse/ditto/thingsearch/service/starter/actors/OperatorSearchMetricsProviderActor.java

```diff
@@ -0,0 +1,398 @@
+                        }
+                        return null;
+                    });
+        }
+    }
+
+    private List<String> ensureThingId(final List<String> fields) {
+        return fields.contains("thingId") ? fields : Stream.concat(fields.stream(), Stream.of("thingId"))
+                .collect(Collectors.toList());
+    }
+
+    private void aggregateByTags(final List<Thing> things,
+            final CustomSearchMetricConfig.FilterConfig filterConfig, final String metricName,
+            final CustomSearchMetricConfig config, final DittoHeaders dittoHeaders) {
+        final List<TagSet> tagSets = things.stream()
+                .map(thing -> TagSet.ofTagCollection(config.getTags().entrySet().stream()
+                        .map(e -> Tag.of(e.getKey(),
+                                resolvePlaceHolder(thing.toJson().asObject(), e.getValue(), filterConfig, dittoHeaders,
+                                        metricName)))
+                        .sorted(Comparator.comparing(Tag::getValue))
+                        .toList())).toList();
+        tagSets.forEach(tagSet -> aggregatedByTagsValues.merge(tagSet, 1.0, Double::sum));
+    }
+
+    private void recordMetric(final String metricName) {
+        aggregatedByTagsValues.forEach(
+                (tags, value) -> metricsGauges.computeIfAbsent(uniqueName(metricName, tags), name -> {
+                            log.info("Initializing custom search metric Gauge for metric <{}> with tags <{}>",
+                                    metricName, tags);
+                            customSearchMetricsGauge.increment();
+                            return Tuple2.apply(KamonGauge.newGauge(metricName)
+                                    .tags(tags), System.currentTimeMillis());
+                        })
+                        ._1().set(value));
+
+        aggregatedByTagsValues.forEach(
+                (tags, value) -> metricsGauges.compute(uniqueName(metricName, tags), (key, tupleValue) -> {
+                            if (tupleValue == null) {
+                                log.info("Initializing custom search metric Gauge for metric <{}> with tags <{}>",
+                                        metricName, tags);
+                                customSearchMetricsGauge.increment();
+                                return Tuple2.apply(KamonGauge.newGauge(metricName)
+                                        .tags(tags), System.currentTimeMillis());
+
+                            } else {
+                                // update the timestamp
+                                log.debug("Updating custom search metric Gauge for metric <{}> with tags <{}>",
+                                        metricName, tags);
+                                return Tuple2.apply(tupleValue._1(), System.currentTimeMillis());
+                            }
+                        })
+                        ._1().set(value));
+
+        aggregatedByTagsValues.clear();
```

Comment:
I don't think this works reliably .. by using the class field `aggregatedByTagsValues`, this field is exposed to access by different threads.

The reason of this is because `recordMetric()` is called within a `Patterns.ask`, in the `whenComplete` - which does not run in the actor's main thread and therefore is not secured by Pekko against multi-threaded access.

So eventually, this will lead to race conditions and very unexpected behavior (e.g. when multiple metrics are gathered and e.g. a lot of results have to be paginated with a cursor over and over again).

I am afraid this has to be changed and re-thought ..